Failed test test_backup_restore_on_cluster/test_cancel_backup.py::test_cancel_restore

Test reports: [report1](https://s3.amazonaws.com/clickhouse-test-reports/72027/bb23ffbcd0e9a3697e676c9f37f0cc9895152eec/integration_tests__tsan__[2_6].html), [report2](https://s3.amazonaws.com/clickhouse-test-reports/71654/0929f66516261dea7b31479b8f5eaac1b4b8e38a/integration_tests__asan__[4_4].html)

```

=================================== FAILURES ===================================
_____________________________ test_cancel_restore ______________________________
[gw3] linux -- Python 3.10.12 /usr/bin/python3

    def test_cancel_restore():
        # Make backup.
        backup_id = get_backup_id_of_successful_backup()
    
        # Cancel restoring.
        with NoTrashChecker() as no_trash_checker:
            print("Will cancel restoring")
            initiator = random_node()
            print(f"Using {get_node_name(initiator)} as initiator")
    
            restore_id = random_id()
            initiator.query(
                f"RESTORE TABLE tbl ON CLUSTER 'cluster' FROM {get_backup_name(backup_id)} SETTINGS id='{restore_id}' ASYNC"
            )
    
            assert get_status(initiator, restore_id=restore_id) == "RESTORING"
            assert get_num_system_processes(initiator, restore_id=restore_id) >= 1
    
            # We shouldn't wait too long here, because otherwise the restore might be completed before we cancel it.
            random_sleep(3)
    
            node_to_cancel, cancel_as_initiator = random.choice(
                [(node1, False), (node2, False), (initiator, True)]
            )
    
            wait_num_system_processes(
                node_to_cancel,
                "1+",
                restore_id=restore_id,
                is_initial_query=cancel_as_initiator,
            )
    
            print(
                f"Cancelling on {'initiator' if cancel_as_initiator else 'node'} {get_node_name(node_to_cancel)}"
            )
    
            # The timeout is 2 seconds here because a restore must be cancelled quickly.
>           kill_query(
                node_to_cancel,
                restore_id=restore_id,
                is_initial_query=cancel_as_initiator,
                timeout=3,
            )

test_backup_restore_on_cluster/test_cancel_backup.py:529: 
```

--- a/tests/integration/test_backup_restore_on_cluster/test_cancel_backup.py
+++ b/tests/integration/test_backup_restore_on_cluster/test_cancel_backup.py
@@ -464,20 +464,21 @@ def test_cancel_backup():
             f"Cancelling on {'initiator' if cancel_as_initiator else 'node'} {get_node_name(node_to_cancel)}"
         )
 
-        # The timeout is 2 seconds here because a backup must be cancelled quickly.
+        time_before_kill_query = time.monotonic()
+
         kill_query(
-            node_to_cancel,
-            backup_id=backup_id,
-            is_initial_query=cancel_as_initiator,
-            timeout=3,
+            node_to_cancel, backup_id=backup_id, is_initial_query=cancel_as_initiator
         )
 
         if cancel_as_initiator:
             assert get_status(initiator, backup_id=backup_id) == "BACKUP_CANCELLED"
-        wait_status(initiator, "BACKUP_CANCELLED", backup_id=backup_id, timeout=3)
+        wait_status(initiator, "BACKUP_CANCELLED", backup_id=backup_id)
+
+        time_to_cancel = time.monotonic() - time_before_kill_query
 
         assert "QUERY_WAS_CANCELLED" in get_error(initiator, backup_id=backup_id)
         assert get_num_system_processes(nodes, backup_id=backup_id) == 0
+        assert time_to_cancel <= 6  # A backup should be cancelled quite quickly.
         no_trash_checker.expect_errors = ["QUERY_WAS_CANCELLED"]
 
 
@@ -518,20 +519,21 @@ def test_cancel_restore():
             f"Cancelling on {'initiator' if cancel_as_initiator else 'node'} {get_node_name(node_to_cancel)}"
         )
 
-        # The timeout is 2 seconds here because a restore must be cancelled quickly.
+        time_before_kill_query = time.monotonic()
+
         kill_query(
-            node_to_cancel,
-            restore_id=restore_id,
-            is_initial_query=cancel_as_initiator,
-            timeout=3,
+            node_to_cancel, restore_id=restore_id, is_initial_query=cancel_as_initiator
         )
 
         if cancel_as_initiator:
             assert get_status(initiator, restore_id=restore_id) == "RESTORE_CANCELLED"
-        wait_status(initiator, "RESTORE_CANCELLED", restore_id=restore_id, timeout=3)
+        wait_status(initiator, "RESTORE_CANCELLED", restore_id=restore_id)
+
+        time_to_cancel = time.monotonic() - time_before_kill_query
 
         assert "QUERY_WAS_CANCELLED" in get_error(initiator, restore_id=restore_id)
         assert get_num_system_processes(nodes, restore_id=restore_id) == 0
+        assert time_to_cancel <= 6  # A restore should be cancelled quite quickly.
         no_trash_checker.expect_errors = ["QUERY_WAS_CANCELLED"]
 
     # Restore successfully.

